internal/imagemgmt/imagemgmt.go

```diff
@@ -0,0 +1,278 @@
+type podmanContainerData struct {
+	ImageID string
+}
+
+// Use a var for syscall.Statfs in order to override it in mock tests
+var syscallStatfs = syscall.Statfs
+
+// CheckDiskUsageAgainstThreshold gets the disk usage for the specified path and checks whether it exceeds the specified threshold
+func (c *ImageMgmtClient) CheckDiskUsageAgainstThreshold(thresholdPercent int) (bool, error) {
+	var stat syscall.Statfs_t
+	if err := syscallStatfs(c.StoragePath, &stat); err != nil {
+		return false, fmt.Errorf("statfs failed for %s: %w", c.StoragePath, err)
+	}
+
+	// Get disk usage percentage
+	usedpct := int(((stat.Blocks - stat.Bavail) * 100) / stat.Blocks)
+	c.Log.Info(fmt.Sprintf("DEBUG: Disk usage of %s is currently %d%%", c.StoragePath, usedpct))
+	return (usedpct > thresholdPercent), nil
+}
+
+// GetInuseImages gets the list of images in use by cri-o or podman containers
+func (c *ImageMgmtClient) GetInuseImages() (inUse []string, rc error) {
+	// Get the list of containers via crictl
+	output, err := c.Executor.Execute("crictl", "ps", "-a", "-o", "json")
+	if err != nil {
+		rc = fmt.Errorf("failed to run crictl ps: %w", err)
+		return
+	}
+
+	var crictlContainers crictlContainersList
+	if err := json.Unmarshal([]byte(output), &crictlContainers); err != nil {
+		rc = fmt.Errorf("unable to parse crictl ps command output: %w", err)
+		return
+	}
+
+	for _, container := range crictlContainers.Containers {
+		inUse = lcautils.AppendToListIfNotExists(inUse, container.Image.Image)
+	}
+
+	// Get the list of containers via podman
+	output, err = c.Executor.Execute("podman", "ps", "-a", "--format", "json")
+	if err != nil {
+		rc = fmt.Errorf("failed to run podman ps: %w", err)
+		return
+	}
+
+	var podmanContainers []podmanContainerData
+	if err := json.Unmarshal([]byte(output), &podmanContainers); err != nil {
+		rc = fmt.Errorf("unable to parse podman ps command output: %w", err)
+		return
+	}
+
+	for _, container := range podmanContainers {
+		inUse = lcautils.AppendToListIfNotExists(inUse, container.ImageID)
+	}
+
+	return
+}
+
+// GetPinnedImages gets the list of pinned images, from crictl
+func (c *ImageMgmtClient) GetPinnedImages() (pinned []string, rc error) {
+	// Get the list of containers via crictl
+	output, err := c.Executor.Execute("crictl", "images", "-o", "json")
+	if err != nil {
+		rc = fmt.Errorf("failed to run crictl images: %w", err)
+		return
+	}
+
+	var crictlImages crictlImagesList
+	if err := json.Unmarshal([]byte(output), &crictlImages); err != nil {
+		rc = fmt.Errorf("unable to parse crictl images command output: %w", err)
+		return
+	}
+
+	for _, image := range crictlImages.Images {
+		if image.Pinned {
+			pinned = lcautils.AppendToListIfNotExists(pinned, image.Id)
+		}
+	}
+
+	return
+}
+
+// isImageUsed compares the specified image info against the list of in-use images to check whether the image ID, tag, or digest is in the in-use list
+func isImageUsed(inUse []string, image imageMgmtImageInfo) bool {
```

Comment:
I wonder if we can just try to remove image and maximum it will not succeed if they are in use?

Reply:
We're also deleting corresponding tags to ensure everything is cleaned up. Ensuring that the image doesn't seem to be in use avoids deleting tags for those images. It also avoids potential unnecessary failures and logs